src/main/java/com/icthh/xm/ms/entity/service/privileges/custom/EntityCustomPrivilegeService.java

```diff
@@ -0,0 +1,198 @@
+                                               String tenantKey,
+                                               String permissionsSpecPath,
+                                               Configuration permissionsSpec) {
+
+        val permissions = updatePermissions(permissionsSpec, specs, tenantKey);
+        String permissionsYml = mapper.writerWithDefaultPrettyPrinter().writeValueAsString(permissions);
+        Configuration configuration = new Configuration(permissionsSpecPath, permissionsYml);
+        if (DigestUtils.sha1Hex(permissionsYml).equals(sha1Hex(permissionsSpec))) {
+            log.info("Permissions configuration not changed");
+            return;
+        }
+        commonConfigRepository.updateConfigFullPath(configuration, sha1Hex(permissionsSpec));
+    }
+
+    private Map<String, ?> updatePermissions(Configuration permissionConfiguration,
+                                             Map<String, TypeSpec> spec,
+                                             String tenantKey) {
+        val permissions = getPermissionsConfig(permissionConfiguration);
+
+        for(CustomPrivilegesExtractor privilegesExtractor: privilegesExtractors) {
+            DefaultPrivilegesValue defaultValue = privilegesExtractor.getDefaultValue();
+            if (defaultValue == NONE) {
+                continue;
+            }
+
+            String sectionName = privilegesExtractor.getSectionName();
+            permissions.computeIfAbsent(sectionName, key -> new TreeMap<>());
+            val permission = permissions.getOrDefault(sectionName, new TreeMap<>());
+
+            Map<String, Role> roles = roleService.getRoles(tenantKey);
+            roles = roles != null ? roles : new HashMap<>();
+            for (String role : roles.keySet()) {
+                permission.computeIfAbsent(role, key -> new TreeSet<>());
+                TreeSet<Permission> rolePermissions = permission.get(role);
+                String privilegePrefix = privilegesExtractor.getPrivilegePrefix();
+
+                privilegesExtractor.toPrivilegesList(spec)
+                                   .stream()
+                                   .map(toPermission(privilegePrefix, defaultValue == ENABLED))
+                                   .filter(not(rolePermissions::contains))
+                                   .forEach(rolePermissions::add);
+            }
+        }
+
+        return permissions;
+    }
+
+    @SneakyThrows
+    private TreeMap<String, TreeMap<String, TreeSet<Permission>>> getPermissionsConfig(
+        Configuration permission) {
+        TreeMap<String, TreeMap<String, TreeSet<Permission>>> permissions = new TreeMap<>();
+        if (isConfigExists(permission)) {
+            permissions = mapper.readValue(permission.getContent(),
+                new TypeReference<TreeMap<String, TreeMap<String, TreeSet<Permission>>>>() {});
+        }
+        return permissions;
+    }
+
+    private Predicate<Permission> not(Predicate<Permission> o) {
```

Comment:
This method is clear utility one which could be helpful somewhere else -> could be `static` and placed somewhere in utility class if such one exist.

Reply:
it's local usage utility method, and used only in this case